### ISSUE TITLE
Problems due overriding configuration that were previously set using env variables

### ISSUE THREAD

**Issue by fefas** [symfony/symfony#22561]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.2.7

Hello,

I don't know if it is a bug/issue or just a question. We are using environment variables to configure the application. For acceptance/feature tests purposes (using `Behat`), we need to switch the database, because we have to arrange the database state for each scenario. In this way, there are the following config files on `app/config` directory:

```yml
# doctrine.yml

doctrine:
  dbal:
    connections:
      default:
        driver: pdo_mysql
        host: "%env(DATABASE_HOST)%"
        dbname: "%env(DATABASE_DBNAME)%"
        user: "%env(DATABASE_USER)%"
        password: "%env(DATABASE_PASSWORD)%"

# ...

# config_dev.yml

imports:
  - resource: doctrine.yml

# ...

# config_behat.yml

imports:
  - resource: doctrine.yml

doctrine:
  dbal:
    connections:
      default:
        host: "%env(DATABASE_TEST_HOST)%"
        dbname: "%env(DATABASE_TEST_DBNAME)%"
        user: "%env(DATABASE_TEST_USER)%"
        password: "%env(DATABASE_TEST_PASSWORD)%"

# ...
```

But the following exception is throwed:

```
[Symfony\Component\DependencyInjection\Exception\EnvParameterException]
  Incompatible use of dynamic environment variables "UPX_DATABASE_HOST", "UPX_DATABASE_DBNAME", "UPX_DATABASE_USER", "UPX_DATABASE_PASSWORD" found in parameters.
```
> It is throwed [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L182).

After reading docs and debugging for a while, we've noted in the code is that _Symfony_ doesn't allow us to override a config value that is an environment variable.

We've noted also that this verification was inserted in PR #19681.

**The question:** Is it Symfony's matter? Or are we missing something here and our configuration is not recommended?

In our humble opinion, problems due to override config values and then not using environment variables previously set should be responsibility of tests that run after application setup (integration, feature, ...).

**Comment by iltar** [comment]
I thought the whole idea of these environment variables was that you can change their values without having to compile the container again. Meaning: for your behat tests, you would change the value of `"%env(DATABASE_HOST)%"` for example, not add a whole new environment variable, or am I missing something here?

**Comment by fefas** [comment]
@iltar .. yes, you are right, but we don't want to change those values in runtime, because those parameters are provided by the environment and the application should just use them.

For us, it doesn't make sense to change the environment variables for other reasons: we want just to switch the database used for the application when running tests that need database setup without change the current state of the application. In this way, we can run all tests anytime without having side effects, what would not happen if those values are changed.

**Comment by Pierstoval** [comment]
Then move these env vars to container parameters and inject the container parameters in your services, it should work at runtime too?

**Comment by xthiago** [comment]
The point is:

I feel that framework should not throw an error if I use or not an environment variable. In the worst case it should display a warning message on cache cleaning or warming up.

We need this exception? Why?

**Comment by Pierstoval** [comment]
Because some options, when used by extensions, imply that you cannot use a dynamic parameter because these options are used before container compilation by extensions or compiler passes.

That's why using a container parameter is a nice option, but I'm not sure whether the cache should be cleared after changing an environment var 😕

**Comment by mattjanssen** [comment]
We just ran into this same issue with a monolog config:

*symfony_prod.yml*
```
monolog:
    handlers:
        main:
            type: loggly
            token: "%env(LOGGLY_TOKEN)%"
```

*symfony_prod_ecs.yml*
```
imports:
    - { resource: symfony_prod.yml }

monolog:
    handlers:
        main:
            type: stream
```

```
[Symfony\Component\DependencyInjection\Exception\EnvParameterException]
Incompatible use of dynamic environment variables "LOGGLY_TOKEN" found in parameters.
```

**Comment by Pierstoval** [comment]
@mattjanssen Have you tried moving the env var to a container parameter?

This would lead to these changes:

```yml
# symfony_prod.yml

**Comment by eXtreme** [comment]
So I am hit by the same problem which mostly happens in my "test" env because I have some special cases (configs) for some bundles/extensions for test env. And overriding bundle configs makes Symfony complain about "unused" vars that was used in "base" configs. As @stof said on slack:

> this is a case where out strict validation (which is here to help you spot mistakes) actually hurts you.

**Comment by nicolas-grekas** [comment]
Should be fixed by #23903
